streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/ClientTagAwareStandbyTaskAssignor.java

```diff
@@ -0,0 +1,289 @@
+
+        final Set<UUID> usedClients = new HashSet<>();
+
+        standbyTaskClientsByTaskLoad.offerAll(clientStates.keySet());
+
+        int numRemainingStandbys = tasksToRemainingStandbys.get(activeTaskId);
+
+        usedClients.add(activeTaskClient);
+
+        final Set<UUID> clientsOnAlreadyUsedTagDimensions = new HashSet<>();
+
+        while (numRemainingStandbys > 0) {
+            clientsOnAlreadyUsedTagDimensions.addAll(
+                findClientsOnUsedTagDimensions(
+                    usedClients,
+                    rackAwareAssignmentTags,
+                    clientStates,
+                    tagEntryToClients,
+                    tagKeyToValues
+                )
+            );
+
+            final UUID polledClient = standbyTaskClientsByTaskLoad.poll(
+                activeTaskId, uuid -> !clientsOnAlreadyUsedTagDimensions.contains(uuid)
+            );
+
+            if (polledClient == null) {
+                break;
+            }
+
+            final ClientState standbyTaskClient = clientStates.get(polledClient);
+
+            if (shouldBalanceLoad(clientStates.values(), standbyTaskClient)) {
+                final Map<String, String> standbyClientTags = standbyTaskClient.clientTags();
+                log.warn("Can't assign {} of {} standby task for task [{}]. " +
+                         "There is not enough capacity on client(s) with {} tag dimensions. " +
+                         "To have a proper distribution of standby tasks across different tag dimensions, " +
+                         "increase the number of application instances with [{}] tag dimensions.",
+                         numRemainingStandbys, numStandbyReplicas, activeTaskId, standbyClientTags, standbyClientTags);
+                break;
+            }
+
+            standbyTaskClient.assignStandby(activeTaskId);
+
+            usedClients.add(polledClient);
+
+            numRemainingStandbys--;
+        }
+
+        return numRemainingStandbys;
+    }
+
+    private static Set<UUID> findClientsOnUsedTagDimensions(final Set<UUID> usedClients,
+                                                            final Set<String> rackAwareAssignmentTags,
+                                                            final Map<UUID, ClientState> clientStates,
+                                                            final Map<TagEntry, Set<UUID>> tagEntryToClients,
+                                                            final Map<String, Set<String>> tagKeyToValues) {
+        final Set<UUID> filteredClients = new HashSet<>();
+
+        for (final UUID usedClientId : usedClients) {
+            final Map<String, String> usedClientTags = clientStates.get(usedClientId).clientTags();
+
+            for (final Entry<String, String> usedClientTagEntry : usedClientTags.entrySet()) {
+                final String tagKey = usedClientTagEntry.getKey();
+
+                if (!rackAwareAssignmentTags.contains(tagKey)) {
+                    continue;
```

Comment:
Is it normal when this happened? Should we do anything to it? Or at least log something here?

Reply:
I think having a warn log is a good call. We can add validation rules (if necessary) when doing last part of this KIP - updating streams configuration.

Reply:
OK